src/main/host/descriptor/tcp.c

```diff
@@ -2329,19 +2334,26 @@ static gssize _tcp_receiveUserData(Transport* transport, gpointer buffer,
 
         /* get the next buffered packet - we'll always need it.
          * this could mark the socket as unreadable if this is its last packet.*/
-        Packet* packet = socket_removeFromInputBuffer((Socket*)tcp);
-        if(!packet) {
+        const Packet* nextPacket = socket_peekNextInPacket((Socket*)tcp);
+        if (!nextPacket) {
             /* no more packets or partial packets */
             break;
         }
 
-        guint packetLength = packet_getPayloadLength(packet);
+        guint packetLength = packet_getPayloadLength(nextPacket);
         copyLength = MIN(packetLength, remaining);
-        bytesCopied = packet_copyPayload(packet, 0, buffer + offset, copyLength);
+        gssize bytesCopied = packet_copyPayload(
+            nextPacket, 0, (PluginVirtualPtr){.val = buffer.val + offset}, copyLength);
+        if (bytesCopied < 0) {
+            // Error writing to PluginVirtualPtr
+            return bytesCopied;
+        }
         totalCopied += bytesCopied;
         remaining -= bytesCopied;
         offset += bytesCopied;
 
+        Packet* packet = socket_removeFromInputBuffer((Socket*)tcp);
```

Comment:
Oh good point.

Otoh I worry that not returning an error here could end up hiding a bug. e.g. in the worst case the plugin has a read loop for which the first "packet size" of the buffer is valid memory, but the rest isn't, and instead of returning an error we cause it to only process one packet of data at a time.

Added a warning for now.

Reply:
I think it's good to have a message here, but having a TCP stream that loses data might be better as an `error()` rather than a `warning()` since your simulation might complete successfully without it being immediately obvious that there were issues. For example applications that detect this error (such as tor using TLS) would close the connection and I think re-open a new one later, and you might not notice this since tor wouldn't crash or exit early.